### ISSUE TITLE
Training doesn't with the last and best weights when We resume training

### ISSUE THREAD

**Issue by coder-rakeshpaul** [ultralytics/ultralytics#5980]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

For testing purpose I trained the detect model on the coco8 dataset with the pretrained weights
the code used for the training was
`
model = YOLO('yolov8n.yaml').load('yolov8n.pt')
results = model.train(data='coco8.yaml', epochs=5 ,imgsz=640, optimizer = 'SGD', momentum= 0.937, lr0= 0.01)
`

and the map values were as follows

![Screenshot 2023-10-28 130427](https://github.com/ultralytics/ultralytics/assets/114281050/20338a8b-0238-41e2-9af5-419856108f89)

after this I used the last.pt to resume and continue the training further and the mAP values drop from the first epoch itself

the code used is 
`model = YOLO('yolov8n.yaml').load('/content/runs/detect/train/weights/last.pt')
results = model.train(data='coco8.yaml', epochs=2 ,imgsz=640, optimizer = 'SGD', momentum= 0.937, lr0= 0.01)`

here are the mAP values
![Screenshot 2023-10-28 130756](https://github.com/ultralytics/ultralytics/assets/114281050/7aaf42f9-ec17-428f-8e61-09fcc8e986ed)

Whereas if the training is continued straight for that many epochs we have some proper values like 

![Screenshot 2023-10-28 131108](https://github.com/ultralytics/ultralytics/assets/114281050/6ef23c2d-3fa0-407b-8f11-ccf333edc397)

### Environment

The environment is of Google colab trained on the T4 gpu

### Minimal Reproducible Example

the initial 5 epochs

`model = YOLO('yolov8n.yaml').load('yolov8n.pt')
results = model.train(data='coco8.yaml', epochs=5 ,imgsz=640, optimizer = 'SGD', momentum= 0.937, lr0= 0.01)`

the next 2 epoch
`model = YOLO('yolov8n.yaml').load('/content/runs/detect/train/weights/last.pt')
results = model.train(data='coco8.yaml', epochs=2 ,imgsz=640, optimizer = 'SGD', momentum= 0.937, lr0= 0.01)`

retraining to check for 7 epochs
`model = YOLO('yolov8n.yaml').load('yolov8n.pt')
results = model.train(data='coco8.yaml', epochs=7 ,imgsz=640, optimizer = 'SGD', momentum= 0.937, lr0= 0.01)`

### Additional

I feel this issue is due to stripping off of the optimizer at the end of training. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by FiksII** [comment]
@coder-rakeshpaul  I think you need to use `resume=True` in train function

**Comment by coder-rakeshpaul** [comment]
@FiksII that doesn't work. It appears that the optimizer state_dicts are stripped off at the end of the training, probably in order to reduce the size of the files, but that has lead to issue. I have managed to prevent the stripping of the optimizer but still haven't figured out how to load them for training

**Comment by glenn-jocher** [comment]
Hello @FiksII,

You are correct in your observation. The optimizer state_dicts are indeed stripped off at the end of the training, as part of the process to reduce the size of the saved weight files. This is intentional, but from your experience, it seems this may be resulting in some issues when resuming training.

In order to load the previous optimizer state for continuing training, you need to ensure that these optimizer states are preserved during the save operation. As you've indicated, you've already taken steps to avoid stripping of the optimizer states.

During the resume operation, along with loading the model weights, it's crucial to load the saved states of the optimizer. This would allow the training to continue from the exact state where it was paused. Importantly, the loading of the optimizer state should be done carefully to ensure that it appropriately corresponds to the loaded model weights.

Thank you for highlighting this, and your efforts in tackling this issue are greatly appreciated. We'll look into this further as it seems to be causing unintentional complications. Be assured that the resolution of this issue will be one of our priorities.

I hope this response proves helpful, and do feel free to reach out if you have more queries or updates.

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul see https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings

**Comment by coder-rakeshpaul** [comment]
Hello @glenn-jocher I appreciate your response could you point out where exactly the weights are loaded?? I am planning to load the optimizer at the very location so as to make the code clean and prevent unnecessary errors. I would also be able to submit a PR in case this is known

I came across this block of code

`    def load(self, weights='yolov8n.pt'):
        """Transfers parameters with matching names and shapes from 'weights' to model."""
        self._check_is_pytorch_model()
        if isinstance(weights, (str, Path)):
            weights, self.ckpt = attempt_load_one_weight(weights)
        self.model.load(weights)
        return self`

Is this the place where I should load the optimizer??

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul your usage is incorrect. A fully trained model can not be resumed. You start a new training using it as pretrained weights, that's all.

Therefore there's no fix or PR required.

**Comment by coder-rakeshpaul** [comment]
@glenn-jocher I understand what you're saying but as mentioned in the initial problem. Why did the mAP value drop so much?

There I have used the pretrained weights of the initial training and nothing else has been modified

so why is there such a huge gap between the mAP values?

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul thanks for your follow-up question. 

When you run training initially for five epochs and then resume it with the model trained on these first five epochs, you're essentially starting a new training cycle. 

The seemingly sudden drop in mAP at the beginning of the new training cycle might initially appear as a discrepancy. However, it's normal and expected for the mAP values to fluctuate, especially in the early stages of training, as the model starts to refit to the dataset with the new training initialization.

During the "new" training with the previously trained weights, the model still goes through an initial adjustment phase, which could result in a temporary drop in mAP values. However, as training progresses, these values should climb back. 

The slight dip in performance in the initial epochs when resuming training is not an indication of an error or an issue with your model. It's simply an artifact of how the training process works.

I hope this helps clarify the situation.

**Comment by coder-rakeshpaul** [comment]
@glenn-jocher thanks for your response while I agree that fluctuations are common in the initial stages of learning it's a small dataset right now(coco8) and a fall of 0.2 is quite significant. Also note that if we just train the initial model for 7 epochs there is no such drop in the value

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul Thanks for your response. I understand your concern. When you resume training with a model that has already been trained on some epochs, the learning rate schedule is reset. This means that the model, even though it starts from a knowledgeable position, begins with a higher learning rate for the resumed training process. 

In the first few epochs of this new training, it may cause significant changes in the weights, resulting in this short-term drop in performance. But usually, it will return to its normal upward trend after several epochs.

Contrarily, when you train the model from scratch for 7 epochs, the learning rate is gradually decreasing from the start, making the loss landscape smoother and hence the drop is not witnessed.

This is typical behaviour and is expected due to the way the learning rate scheduling works. I hope that clarifies your query.

**Comment by coder-rakeshpaul** [comment]
@glenn-jocher thank you for your answer. Nevertheless I still wish to save the optimizer state_dict for my current project. 

I have managed to prevent the stripping of the optimizer state_dicts.

And now want to figure out how to load them when starting the training again (PS: I'm not resuming the training I want to start the training all over again but with these new weights and saved optimizer state_dicts). 

Any suggestions on how that can be accomplished??

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul, 

Indeed, you can continue training with the saved weights and optimizer state_dicts. 

To achieve this, you need to load both the model weights and the optimizer state_dict at the point where you initialize your model and optimizer for the new training session. On loading, make certain that the state_dict you're loading matches the architecture of your model and optimizer. The important thing is to ensure that the loaded optimizer state_dict corresponds to the loaded weights.

This should give you the behavior you're expecting -- restarting a new training session with the specified weights and optimizer states. 

Remember that this won't be a straight continuation of the previous training session as you're resetting the learning rate schedule, batch statistics etc. Hence there still could be dips in performance in early epochs which is normal and expected. 

Happy coding!
